sdk/python/runtime/main.go

```diff
@@ -24,44 +35,57 @@ const (
 //go:embed scripts/runtime.py
 var runtimeTmpl string
 
-func (m *PythonSdk) ModuleRuntime(modSource *Directory, subPath string, introspectionJson string) *Container {
-	return m.CodegenBase(modSource, subPath, introspectionJson).
+func (m *PythonSdk) ModuleRuntime(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*Container, error) {
+	ctr, err := m.CodegenBase(ctx, modSource, introspectionJson)
+	if err != nil {
+		return nil, err
+	}
+
+	return ctr.
 		WithExec([]string{"python", "-m", "pip", "install", "."}).
-		WithWorkdir(ModSourceDirPath).
 		WithNewFile(RuntimeExecutablePath, ContainerWithNewFileOpts{
 			Contents:    runtimeTmpl,
 			Permissions: 0755,
 		}).
-		WithEntrypoint([]string{RuntimeExecutablePath})
+		WithEntrypoint([]string{RuntimeExecutablePath}), nil
 }
 
-func (m *PythonSdk) Codegen(modSource *Directory, subPath string, introspectionJson string) *GeneratedCode {
-	ctr := m.CodegenBase(modSource, subPath, introspectionJson)
+func (m *PythonSdk) Codegen(ctx context.Context, modSource *ModuleSource, introspectionJson string) (*GeneratedCode, error) {
+	ctr, err := m.CodegenBase(ctx, modSource, introspectionJson)
+	if err != nil {
+		return nil, err
+	}
+
 	ctr = ctr.WithDirectory(genDir, ctr.Directory(sdkSrc), ContainerWithDirectoryOpts{
 		Exclude: []string{
 			"**/__pycache__",
 		},
 	})
 
 	modified := ctr.Directory(ModSourceDirPath)
-	diff := modSource.Diff(modified).Directory(subPath)
+	diff := modSource.RootDirectory().Diff(modified)
 
 	return dag.GeneratedCode(diff).
-		WithVCSIgnoredPaths([]string{
-			genDir,
-		})
+		WithVCSGeneratedPaths([]string{
+			genDir + "/**",
+		}), nil
 }
 
-func (m *PythonSdk) CodegenBase(modSource *Directory, subPath string, introspectionJson string) *Container {
+func (m *PythonSdk) CodegenBase(ctx context.Context, modSource *ModuleSource, introspectionJson string) (*Container, error) {
+	subPath, err := modSource.Subpath(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not load module config: %v", err)
+	}
+
 	return m.Base("").
-		WithDirectory(sdkSrc, dag.Host().Directory(root(), HostDirectoryOpts{
-			Exclude: []string{"runtime"},
-		})).
-		WithMountedDirectory("/opt", dag.Host().Directory(root(), HostDirectoryOpts{
-			Include: []string{"runtime/template"},
-		})).
+		WithMountedDirectory(sdkSrc, m.SDKSourceDir).
+		WithMountedDirectory("/opt", dag.CurrentModule().Source().Directory("./template")).
```

Comment:
Yeah, this gets confusing in the specific case of "SDK-as-module", which gets all "meta", but:
1. `dag.CurrentModule().Source().Directory("./template")` is reading a file from the *python SDK as a module* source dir. So it's the dir `./sdk/python/runtime/template` in our repo (where `./sdk/python/runtime` is the source dir).
2. `modSource.Directory("./template")` would be referring to the source dir of the *module that the python sdk is building* (i.e. some module written in the python sdk that's being built now)

Thankfully, this distinction only ever matters when modules accept other modules as arguments, which for now is limited to SDKs (though there's all kinds of interesting use cases for this in general). So users shouldn't have to think about it typically.

Reply:
Sorry, I had no business reviewing while tired. 😇 I made the same mistake in confusing this in other comments so desregard them as well.

I actually remembered this might be the case after getting away from the computer and my head hit the pillow.